File: src/main/java/com/code/tama/triggerapi/rendering/RenderUtils.java
/* (C) TAMA Studios 2025 */
package com.code.tama.triggerapi.rendering;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.world.phys.AABB;

import com.code.tama.triggerapi.NativeLoader;

/**
 * Rendering math utilities backed by native Rust for batch operations.
 *
 * The pattern here is always: Java handles MC objects and GL calls,
 * Rust handles bulk math on flat float arrays. Never call the native
 * methods for single operations, the JNI overhead isn't worth it.
 * Use them when you have 10+ objects to process per frame.
 */
public class RenderUtils {

    static { NativeLoader.load("tts_native"); }

    // -- Native (Rust) --------------------------------------------------------

    /**
     * Frustum-culls a batch of AABBs against 6 frustum planes.
     *
     * @param aabbs flat array of [minX,minY,minZ,maxX,maxY,maxZ, ...], one entry per AABB
     * @param frustumPlanes 6 planes × 4 floats [nx,ny,nz,d] = 24 floats total
     * @return indices of AABBs that passed (are potentially visible)
     */
    public static native int[] frustumCull(float[] aabbs, float[] frustumPlanes);

    /**
     * Generates all 24 line-segment vertices (bottom face, top face, 4 vertical edges)
     * for N AABBs at once. Output per vertex: [x, y, z, nx, ny, nz], 144 floats per AABB.
     * Feed the result directly into a VertexConsumer; no per-vertex math needed in Java.
     *
     * @param aabbs flat [minX,minY,minZ,maxX,maxY,maxZ, ...]
     */
    public static native float[] buildAABBVertices(float[] aabbs);

    /**
     * Sorts N AABBs by distance from a camera point.
     *
     * @param aabbs flat [minX,minY,minZ,maxX,maxY,maxZ, ...]
     * @param reverse 1 = farthest first (use for transparent geometry), 0 = nearest first
     * @return sorted indices into the original aabbs array
     */
    public static native int[] sortByDistance(float[] aabbs, float cx, float cy, float cz, int reverse);

    /**
     * Packs ARGB float components (each 0.0–1.0) into a single int (0xAARRGGBB).
     * Marginally faster than doing it in Java for tight loops.
     */
    public static native int packARGB(float a, float r, float g, float b);

    /**
     * Transforms N xyz points by a 4×4 column-major matrix (same layout as JOML/GL).
     *
     * @param points flat [x,y,z, x,y,z, ...]
     * @param matrix 16 floats, column-major
     * @return transformed flat [x',y',z', ...]
     */
    public static native float[] transformPoints(float[] points, float[] matrix);

    /**
     * Generates vertices for a wireframe sphere as line segment pairs.
     * Returns flat [x,y,z, x,y,z, ...], each consecutive pair is one line.
     *
     * @param stacks latitude resolution (8–16 is plenty for debug rendering)
     * @param slices longitude resolution
     */
    public static native float[] buildSphereVertices(float cx, float cy, float cz, float radius, int stacks, int slices);

    /**
     * Computes normalised direction vectors for a batch of line segments.
     * Useful when you need to supply normals for RenderType.lines() separately.
     *
     * @param lines flat [x0,y0,z0,x1,y1,z1, ...]
     * @return flat [nx,ny,nz, ...], one normal per line
     */
    public static native float[] computeLineNormals(float[] lines);

    /**
     * Inflates (expands) a batch of AABBs by a fixed amount on all sides.
     *
     * @param aabbs flat [minX,minY,minZ,maxX,maxY,maxZ, ...]
     * @param amount how much to expand on each side (negative = shrink)
     */
    public static native float[] inflateAABBs(float[] aabbs, float amount);

    // -- Java helpers, convenience wrappers that do the MC↔float[] conversion --

    /**
     * Packs a list of AABBs into the flat float array format Rust expects.
     * Call once, reuse the result across frustumCull / buildAABBVertices / sortByDistance.
     */
    public static float[] packAABBs(AABB... boxes) {
        float[] out = new float[boxes.length * 6];
        for (int i = 0; i < boxes.length; i++) {
            AABB b = boxes[i];
            int base = i * 6;
            out[base]   = (float) b.minX;  out[base+1] = (float) b.minY;  out[base+2] = (float) b.minZ;
            out[base+3] = (float) b.maxX;  out[base+4] = (float) b.maxY;  out[base+5] = (float) b.maxZ;
        }
        return out;
    }

    /**
     * Submits pre-built AABB vertices (from buildAABBVertices) to a VertexConsumer.
     * The vertex data already has normals baked in, just pump them through.
     *
     * @param vertexData output of buildAABBVertices()
     * @param pose       current PoseStack.last()
     * @param consumer   the VertexConsumer for RenderType.lines()
     * @param r,g,b,a    line colour
     */
    public static void submitAABBVertices(
            float[] vertexData,
            PoseStack.Pose pose,
            VertexConsumer consumer,
            float r, float g, float b, float a) {

        // Each vertex is 6 floats: [x, y, z, nx, ny, nz]
        for (int i = 0; i < vertexData.length; i += 6) {
            consumer.vertex(pose.pose(),   vertexData[i],   vertexData[i+1], vertexData[i+2])
                    .color(r, g, b, a)
                    .normal(pose.normal(), vertexData[i+3], vertexData[i+4], vertexData[i+5])
                    .endVertex();
        }
    }

    /**
     * Submits sphere vertices (from buildSphereVertices) to a VertexConsumer.
     * Computes normals on the fly since spheres need them per-segment.
     */
    public static void submitSphereVertices(
            float[] vertexData,
            PoseStack.Pose pose,
            VertexConsumer consumer,
            float r, float g, float b, float a) {

        float[] normals = computeLineNormals(vertexData);

        // vertex data: [x0,y0,z0, x1,y1,z1, ...], pairs of endpoints
        // normals:     [nx,ny,nz, ...], one per pair
        for (int i = 0, n = 0; i < vertexData.length; i += 6, n += 3) {
            float nx = normals[n], ny = normals[n+1], nz = normals[n+2];
            consumer.vertex(pose.pose(), vertexData[i],   vertexData[i+1], vertexData[i+2])
                    .color(r, g, b, a)
                    .normal(pose.normal(), nx, ny, nz)
                    .endVertex();
            consumer.vertex(pose.pose(), vertexData[i+3], vertexData[i+4], vertexData[i+5])
                    .color(r, g, b, a)
                    .normal(pose.normal(), nx, ny, nz)
                    .endVertex();
        }
    }

    // -- Usage example ---------------------------------------------------------
    //
    // Rendering 500 entity hitboxes each frame, old way (all in Java):
    //
    //   for (Entity e : entities) {
    //       renderAABBLines(poseStack, consumer, e.getBoundingBox(), r, g, b, a);
    //   }
    //
    // New way, Rust does the bulk math, Java just submits to the GPU:
    //
    //   // Pack all AABBs once
    //   AABB[] boxes = entities.stream().map(Entity::getBoundingBox).toArray(AABB[]::new);
    //   float[] packed = RenderUtils.packAABBs(boxes);
    //
    //   // Optional: cull against frustum first
    //   int[] visible = RenderUtils.frustumCull(packed, frustumPlanes);
    //   float[] culledPacked = filterByIndices(packed, visible); // helper below
    //
    //   // Optional: sort back-to-front for transparent lines
    //   // int[] sorted = RenderUtils.sortByDistance(culledPacked, camX, camY, camZ, 1);
    //
    //   // Build all vertices in one native call
    //   float[] verts = RenderUtils.buildAABBVertices(culledPacked);
    //
    //   // Submit to GPU, just buffer writes, no more per-edge sqrt
    //   RenderUtils.submitAABBVertices(verts, poseStack.last(), consumer, r, g, b, a);

    /** Filters a packed AABB array down to only the indices returned by frustumCull. */
    public static float[] filterByIndices(float[] aabbs, int[] indices) {
        float[] out = new float[indices.length * 6];
        for (int i = 0; i < indices.length; i++) {
            System.arraycopy(aabbs, indices[i] * 6, out, i * 6, 6);
        }
        return out;
    }
}

File: src/native/src/render_utils.rs
use jni::JNIEnv;
use jni::objects::{JClass, JFloatArray};
use jni::sys::{jfloat, jfloatArray, jint, jintArray};

// Helper: copy a Java float[] into a Rust Vec<f32>
fn get_floats(env: &mut JNIEnv, arr: &JFloatArray) -> Option<Vec<f32>> {
    let len = env.get_array_length(arr).ok()? as usize;
    let mut buf = vec![0f32; len];
    env.get_float_array_region(arr, 0, &mut buf).ok()?;
    Some(buf)
}

// Helper: copy a Rust &[f32] into a new Java float[]
fn to_jfloat_array(env: &mut JNIEnv, data: &[f32]) -> jfloatArray {
    match env.new_float_array(data.len() as i32) {
        Ok(arr) => {
            let _ = env.set_float_array_region(&arr, 0, data);
            arr.into_raw()
        }
        Err(_) => std::ptr::null_mut(),
    }
}

// Helper: copy a Rust &[i32] into a new Java int[]
fn to_jint_array(env: &mut JNIEnv, data: &[i32]) -> jintArray {
    match env.new_int_array(data.len() as i32) {
        Ok(arr) => {
            let _ = env.set_int_array_region(&arr, 0, data);
            arr.into_raw()
        }
        Err(_) => std::ptr::null_mut(),
    }
}

// ── Frustum Culling ───────────────────────────────────────────────────────────
// aabbs: flat [minX,minY,minZ,maxX,maxY,maxZ, ...]
// frustumPlanes: 6 planes × 4 floats [nx,ny,nz,d]
// Returns indices of visible AABBs.
// Java: native int[] frustumCull(float[] aabbs, float[] frustumPlanes);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_frustumCull(
    mut env: JNIEnv,
    _class: JClass,
    aabbs: JFloatArray,
    frustum_planes: JFloatArray,
) -> jintArray {
    let aabb_data  = match get_floats(&mut env, &aabbs)         { Some(v) => v, None => return std::ptr::null_mut() };
    let plane_data = match get_floats(&mut env, &frustum_planes) { Some(v) => v, None => return std::ptr::null_mut() };

    let count = aabb_data.len() / 6;
    let mut visible: Vec<i32> = Vec::with_capacity(count);

    for i in 0..count {
        let b = i * 6;
        let min_x = aabb_data[b];     let min_y = aabb_data[b+1]; let min_z = aabb_data[b+2];
        let max_x = aabb_data[b+3];   let max_y = aabb_data[b+4]; let max_z = aabb_data[b+5];

        let mut inside = true;
        for p in 0..6 {
            let pb = p * 4;
            let nx = plane_data[pb]; let ny = plane_data[pb+1];
            let nz = plane_data[pb+2]; let d = plane_data[pb+3];

            let px = if nx >= 0.0 { max_x } else { min_x };
            let py = if ny >= 0.0 { max_y } else { min_y };
            let pz = if nz >= 0.0 { max_z } else { min_z };

            if nx*px + ny*py + nz*pz + d < 0.0 {
                inside = false;
                break;
            }
        }

        if inside { visible.push(i as i32); }
    }

    to_jint_array(&mut env, &visible)
}

// ── Batch AABB Vertex Generation ──────────────────────────────────────────────
// Output per vertex: [x, y, z, nx, ny, nz],  24 vertices per AABB = 144 floats per box.
// Java unpacks and feeds directly to VertexConsumer.
// Java: native float[] buildAABBVertices(float[] aabbs);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_buildAABBVertices(
    mut env: JNIEnv,
    _class: JClass,
    aabbs: JFloatArray,
) -> jfloatArray {
    let data = match get_floats(&mut env, &aabbs) { Some(v) => v, None => return std::ptr::null_mut() };

    let count = data.len() / 6;
    let mut out: Vec<f32> = Vec::with_capacity(count * 144);

    for i in 0..count {
        let b = i * 6;
        let x0 = data[b];   let y0 = data[b+1]; let z0 = data[b+2];
        let x1 = data[b+3]; let y1 = data[b+4]; let z1 = data[b+5];

        // Bottom face
        push_line(&mut out, x0,y0,z0, x1,y0,z0);
        push_line(&mut out, x1,y0,z0, x1,y0,z1);
        push_line(&mut out, x1,y0,z1, x0,y0,z1);
        push_line(&mut out, x0,y0,z1, x0,y0,z0);
        // Top face
        push_line(&mut out, x0,y1,z0, x1,y1,z0);
        push_line(&mut out, x1,y1,z0, x1,y1,z1);
        push_line(&mut out, x1,y1,z1, x0,y1,z1);
        push_line(&mut out, x0,y1,z1, x0,y1,z0);
        // Vertical edges
        push_line(&mut out, x0,y0,z0, x0,y1,z0);
        push_line(&mut out, x1,y0,z0, x1,y1,z0);
        push_line(&mut out, x1,y0,z1, x1,y1,z1);
        push_line(&mut out, x0,y0,z1, x0,y1,z1);
    }

    to_jfloat_array(&mut env, &out)
}

// ── Distance Sorting ──────────────────────────────────────────────────────────
// Sorts N AABBs by distance from a point. reverse=1 → farthest first (transparency).
// Java: native int[] sortByDistance(float[] aabbs, float cx, float cy, float cz, int reverse);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_sortByDistance(
    mut env: JNIEnv,
    _class: JClass,
    aabbs: JFloatArray,
    cx: jfloat, cy: jfloat, cz: jfloat,
    reverse: jint,
) -> jintArray {
    let data = match get_floats(&mut env, &aabbs) { Some(v) => v, None => return std::ptr::null_mut() };

    let count = data.len() / 6;
    let mut indexed: Vec<(i32, f32)> = (0..count as i32).map(|i| {
        let b = i as usize * 6;
        let mx = (data[b] + data[b+3]) * 0.5 - cx;
        let my = (data[b+1] + data[b+4]) * 0.5 - cy;
        let mz = (data[b+2] + data[b+5]) * 0.5 - cz;
        (i, mx*mx + my*my + mz*mz)
    }).collect();

    if reverse != 0 {
        indexed.sort_unstable_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    } else {
        indexed.sort_unstable_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));
    }

    let indices: Vec<i32> = indexed.into_iter().map(|(i, _)| i).collect();
    to_jint_array(&mut env, &indices)
}

// ── Color Packing ─────────────────────────────────────────────────────────────
// Java: native int packARGB(float a, float r, float g, float b);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_packARGB(
    _env: JNIEnv,
    _class: JClass,
    a: jfloat, r: jfloat, g: jfloat, b: jfloat,
) -> jint {
    let ai = (a.clamp(0.0,1.0) * 255.0) as u32;
    let ri = (r.clamp(0.0,1.0) * 255.0) as u32;
    let gi = (g.clamp(0.0,1.0) * 255.0) as u32;
    let bi = (b.clamp(0.0,1.0) * 255.0) as u32;
    ((ai << 24) | (ri << 16) | (gi << 8) | bi) as i32
}

// ── Batch Point Transform ─────────────────────────────────────────────────────
// Transforms N xyz points by a 4×4 column-major matrix (same as JOML/GL).
// Java: native float[] transformPoints(float[] points, float[] matrix);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_transformPoints(
    mut env: JNIEnv,
    _class: JClass,
    points: JFloatArray,
    matrix: JFloatArray,
) -> jfloatArray {
    let pts = match get_floats(&mut env, &points) { Some(v) => v, None => return std::ptr::null_mut() };
    let mx  = match get_floats(&mut env, &matrix)  { Some(v) => v, None => return std::ptr::null_mut() };
    if mx.len() < 16 { return std::ptr::null_mut(); }

    let count = pts.len() / 3;
    let mut out = Vec::with_capacity(pts.len());

    for i in 0..count {
        let x = pts[i*3]; let y = pts[i*3+1]; let z = pts[i*3+2];
        out.push(mx[0]*x + mx[4]*y + mx[8]*z  + mx[12]);
        out.push(mx[1]*x + mx[5]*y + mx[9]*z  + mx[13]);
        out.push(mx[2]*x + mx[6]*y + mx[10]*z + mx[14]);
    }

    to_jfloat_array(&mut env, &out)
}

// ── Sphere Vertex Generation ──────────────────────────────────────────────────
// Returns flat [x,y,z, x,y,z, ...] line segment pairs for a wireframe sphere.
// Java: native float[] buildSphereVertices(float cx, float cy, float cz, float radius, int stacks, int slices);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_buildSphereVertices(
    mut env: JNIEnv,
    _class: JClass,
    cx: jfloat, cy: jfloat, cz: jfloat,
    radius: jfloat,
    stacks: jint, slices: jint,
) -> jfloatArray {
    let stacks = stacks.max(2) as usize;
    let slices = slices.max(3) as usize;
    let mut out: Vec<f32> = Vec::with_capacity(stacks * slices * 12);
    use std::f32::consts::PI;

    for i in 0..=stacks {
        let phi = PI * i as f32 / stacks as f32;
        let y = cy + radius * phi.cos();
        let r = radius * phi.sin();
        for j in 0..slices {
            let t0 = 2.0 * PI * j as f32 / slices as f32;
            let t1 = 2.0 * PI * (j+1) as f32 / slices as f32;
            out.extend_from_slice(&[cx+r*t0.cos(), y, cz+r*t0.sin(),
                                    cx+r*t1.cos(), y, cz+r*t1.sin()]);
        }
    }
    for j in 0..slices {
        let t = 2.0 * PI * j as f32 / slices as f32;
        let (ct, st) = (t.cos(), t.sin());
        for i in 0..stacks {
            let p0 = PI * i as f32 / stacks as f32;
            let p1 = PI * (i+1) as f32 / stacks as f32;
            out.extend_from_slice(&[cx+radius*p0.sin()*ct, cy+radius*p0.cos(), cz+radius*p0.sin()*st,
                                    cx+radius*p1.sin()*ct, cy+radius*p1.cos(), cz+radius*p1.sin()*st]);
        }
    }

    to_jfloat_array(&mut env, &out)
}

// ── Line Normal Batch ─────────────────────────────────────────────────────────
// Given line endpoint pairs [x0,y0,z0,x1,y1,z1, ...] returns [nx,ny,nz, ...].
// Java: native float[] computeLineNormals(float[] lines);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_computeLineNormals(
    mut env: JNIEnv,
    _class: JClass,
    lines: JFloatArray,
) -> jfloatArray {
    let data = match get_floats(&mut env, &lines) { Some(v) => v, None => return std::ptr::null_mut() };

    let count = data.len() / 6;
    let mut out = Vec::with_capacity(count * 3);

    for i in 0..count {
        let b = i * 6;
        let dx = data[b+3]-data[b]; let dy = data[b+4]-data[b+1]; let dz = data[b+5]-data[b+2];
        let len = (dx*dx + dy*dy + dz*dz).sqrt();
        if len > 0.0 { out.extend_from_slice(&[dx/len, dy/len, dz/len]); }
        else         { out.extend_from_slice(&[0.0, 1.0, 0.0]); }
    }

    to_jfloat_array(&mut env, &out)
}

// ── AABB Inflate ──────────────────────────────────────────────────────────────
// Java: native float[] inflateAABBs(float[] aabbs, float amount);
#[no_mangle]
pub extern "system" fn Java_com_code_tama_triggerapi_helpers_RenderUtils_inflateAABBs(
    mut env: JNIEnv,
    _class: JClass,
    aabbs: JFloatArray,
    amount: jfloat,
) -> jfloatArray {
    let data = match get_floats(&mut env, &aabbs) { Some(v) => v, None => return std::ptr::null_mut() };

    let mut out: Vec<f32> = Vec::with_capacity(data.len());
    for i in 0..data.len()/6 {
        let b = i * 6;
        out.extend_from_slice(&[
            data[b]-amount,   data[b+1]-amount, data[b+2]-amount,
            data[b+3]+amount, data[b+4]+amount, data[b+5]+amount,
        ]);
    }

    to_jfloat_array(&mut env, &out)
}

// ── internal ──────────────────────────────────────────────────────────────────

#[inline(always)]
fn push_line(out: &mut Vec<f32>, x0:f32,y0:f32,z0:f32, x1:f32,y1:f32,z1:f32) {
    let dx = x1-x0; let dy = y1-y0; let dz = z1-z0;
    let len = (dx*dx + dy*dy + dz*dz).sqrt();
    let (nx,ny,nz) = if len > 0.0 { (dx/len, dy/len, dz/len) } else { (0.0,1.0,0.0) };
    out.extend_from_slice(&[x0,y0,z0,nx,ny,nz, x1,y1,z1,nx,ny,nz]);
}